### PR TITLE
error when passing a non-identifier to connect

### LINKED ISSUE #764: improve error message when passing a non-identifier to start a connection
we allow users to pass a connection variable to connect to a db. example:

```python
%load_ext sql

import duckdb
conn = duckdb.connect()

%sql conn
```

however, in some cases, the connection object might be part of a dictionary, list or object. examples:

```python
%sql object.property
%sql dictionary["key"]
%sql some_list[0]
```

Trying any of the above throws an error because our parser thinks it's SQL code. Instead, we should provide a more informative error:

```
"object.property" is not a valid connection identifier. Please pass the variable's name directly, as passing object attributes, dictionaries `some_dict[key]` or lists `some_list[0]` won't work. Example: `%sql connection`
```

so we need a way to identity such cases, we can use the following heuristics:

- the magic gets a single argument (no spaces between characters) - because valid SQL code will always have spaces
- we can check if the argument is not a valid identifier using `argument.isidentifier()` but it is valid Python code (using ast.parse)

if those conditions hold, then we can raise the error.

### PR DESCRIPTION
**Pull request by bbeat2782**
## Describe your changes

When a connection is pass through dictionary, list, or object, it raises an error with a more detailed message.

For instance, when `%sql some_list[0]` is passed to connect, it raises the following error.
```
UsageError: 'some_list[0]' is not a valid connection identifier. Please pass the variable's name directly, as passing object attributes, dictionaries or lists won't work.
```

## Issue number

Closes #764 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--821.org.readthedocs.build/en/821/

### CODE REVIEW

File: src/sql/magic.py
Comment (edublancas) [inline src/sql/magic.py:506]: I can't remember exactly where's the logic that parses this:

```
import duckdb
some_conn = duckdb.connect()

%sql some_conn
```

(looks up `some_conn` in the user's namespace and use it as a connection)

I think it's in the command.py file. it'll be better to move this logic to that function because so all the logic is in one place

File: src/sql/util.py
Comment (edublancas) [inline src/sql/util.py:551]: this logic is hard to read

`command_sql_first_line` isn't the first line, these are the first line tokens 

I think if you move this to wherever we're parsing the existing logic, you'll find similar logic already, and it'll simplify this

File: src/sql/util.py
Comment (edublancas) [inline src/sql/util.py:560]: this part is unclear, you can abstract the ast.parse into a function like `is_valid_python_code`, because having the ast.parse line and a raise  inside a try statement is confusing

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/magic.py
Comment (bbeat2782) [reply src/sql/magic.py]: [fixed](https://github.com/ploomber/jupysql/pull/821/files#diff-769097f5e36be57ddec10c78df0900db8c1e254f5485bc94367aa7fcb07991c5R72) : Moved to `SQLCommand` in `command.py` where it determines a connection

File: src/sql/util.py
Comment (bbeat2782) [reply src/sql/util.py]: [fixed](https://github.com/ploomber/jupysql/pull/821/files#diff-653c15e6161c97f356be04e072838ed15014af275c06019a7692ab727fa44b97R543-R556) : created `is_valid_python_code` function to return `True/False` based on `ast.parse(code)` and raised the `exceptions.UsageError` inside a if statement

File: src/sql/util.py
Comment (bbeat2782) [reply src/sql/util.py]: Only test `isidentifier()` and `is_valid_python_code` in `validate_nonidentifier_connection` : [fixed](https://github.com/ploomber/jupysql/pull/821/files#diff-653c15e6161c97f356be04e072838ed15014af275c06019a7692ab727fa44b97R543)

Simplified if conditions by using what's available in `command.py` : [fixed](https://github.com/ploomber/jupysql/pull/821/files#diff-769097f5e36be57ddec10c78df0900db8c1e254f5485bc94367aa7fcb07991c5R65-R69)

File: src/sql/command.py
Comment (edublancas) [inline src/sql/command.py:63]: unclear about this part. is this some new behavior? or is it refactoring of existing one?

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/command.py
Comment (bbeat2782) [reply src/sql/command.py]: Finding the variable from user_ns exists in the later part of the code with `self._var_expand`: [here](https://github.com/ploomber/jupysql/blob/master/src/sql/command.py#L70)

So I relocated the code for checking nonidentifier after `_var_expand`. Now, the condition for checking `{{` and `}}` and `[` and `]` (for the deprecated feature) are gone. : [fixed](https://github.com/ploomber/jupysql/pull/821/files#diff-769097f5e36be57ddec10c78df0900db8c1e254f5485bc94367aa7fcb07991c5R85-R91)

**Review by edublancas — DISMISSED** [review]

**Comment by edublancas** [comment]
@bbeat2782 there were some merge conflicts in the changelog, but shouldn't affect the CI so please ping me once they pass so I can merge

**Comment by bbeat2782** [comment]
@edublancas It's weird that I'm not seeing any merge conflict

**Comment by edublancas** [comment]
@bbeat2782 ah, yes. i fixed them and was just waiting for the CI (it was a changelog minor issue)

**Pull request merged**
